File: packages_py/fetch_auth_config/src/fetch_auth_config/resolvers/basic.py
from ..types.auth_config import AuthConfig, AuthResolutionMeta
from ..types.auth_type import AuthType
from ..types.token_resolver import TokenResolverType
from ..errors import MissingCredentialError
from ..resolution.config_extractor import ProviderEnvMappings
from ..resolution.env_resolver import resolve_env_var_chain

def resolve_basic_auth(
    provider_name: str,
    env_mappings: ProviderEnvMappings,
    auth_type: AuthType
) -> AuthConfig:
    resolved_from = {}
    
    # Username/Email resolution
    username = None
    email = None
    
    if auth_type in [AuthType.BASIC, AuthType.BASIC_TOKEN, AuthType.BASIC_EMAIL, AuthType.BASIC_EMAIL_TOKEN]:
        # Needs username or email
        if auth_type in [AuthType.BASIC_EMAIL, AuthType.BASIC_EMAIL_TOKEN]:
             res = resolve_env_var_chain(primary=env_mappings.email.primary, overwrite=env_mappings.email.overwrite)
             if not res.value: raise MissingCredentialError(provider_name, 'email', res.tried)
             email = res.value
             resolved_from['email'] = res.source
        else:
             res = resolve_env_var_chain(primary=env_mappings.username.primary, overwrite=env_mappings.username.overwrite)
             if not res.value: raise MissingCredentialError(provider_name, 'username', res.tried)
             username = res.value
             resolved_from['username'] = res.source

    # Password/Token resolution
    password = None
    token = None
    
    if auth_type in [AuthType.BASIC, AuthType.BASIC_EMAIL]:
        res = resolve_env_var_chain(primary=env_mappings.password.primary, overwrite=env_mappings.password.overwrite)
        if not res.value: raise MissingCredentialError(provider_name, 'password', res.tried)
        password = res.value
        resolved_from['password'] = res.source
    elif auth_type in [AuthType.BASIC_TOKEN, AuthType.BASIC_EMAIL_TOKEN]:
        res = resolve_env_var_chain(primary=env_mappings.api_key.primary, overwrite=env_mappings.api_key.overwrite)
        if not res.value: raise MissingCredentialError(provider_name, 'token', res.tried)
        token = res.value
        resolved_from['token'] = res.source

    return AuthConfig(
        type=auth_type,
        provider_name=provider_name,
        username=username,
        email=email,
        password=password,
        token=token,
        header_name="Authorization",
        resolution=AuthResolutionMeta(
            resolved_from=resolved_from,
            token_resolver=TokenResolverType.STATIC,
            is_placeholder=False
        )
    )

File: packages_mjs/fetch-auth-config/src/resolvers/basic.ts
import { AuthConfig, AuthResolutionMeta } from '../types/auth-config.js';
import { AuthType } from '../types/auth-type.js';
import { TokenResolverType } from '../types/token-resolver.js';
import { MissingCredentialError } from '../errors.js';
import { ProviderEnvMappings } from '../resolution/config-extractor.js';
import { resolveEnvVarChain } from '../resolution/env-resolver.js';

export function resolveBasicAuth(
    providerName: string,
    envMappings: ProviderEnvMappings,
    authType: AuthType
): AuthConfig {
    const resolvedFrom: Record<string, string> = {};

    let username: string | undefined;
    let email: string | undefined;
    let password: string | undefined;
    let token: string | undefined;

    // Username/Email
    if ([AuthType.BASIC_EMAIL, AuthType.BASIC_EMAIL_TOKEN].includes(authType)) {
        const res = resolveEnvVarChain(envMappings.email);
        if (!res.value) throw new MissingCredentialError(providerName, 'email', res.tried);
        email = res.value;
        resolvedFrom['email'] = res.source!;
    } else {
        const res = resolveEnvVarChain(envMappings.username);
        if (!res.value) throw new MissingCredentialError(providerName, 'username', res.tried);
        username = res.value;
        resolvedFrom['username'] = res.source!;
    }

    // Password/Token
    if ([AuthType.BASIC, AuthType.BASIC_EMAIL].includes(authType)) {
        let res = resolveEnvVarChain(envMappings.password);
        if (!res.value) {
            // Fallback: try using apiKey as password (common pattern where endpoint_api_key is used for password)
            const tokenRes = resolveEnvVarChain(envMappings.apiKey);
            if (tokenRes.value) {
                res = tokenRes;
            } else {
                throw new MissingCredentialError(providerName, 'password', res.tried);
            }
        }
        password = res.value ?? undefined;
        resolvedFrom['password'] = res.source!;
    } else {
        const res = resolveEnvVarChain(envMappings.apiKey);
        if (!res.value) throw new MissingCredentialError(providerName, 'token', res.tried);
        token = res.value;
        resolvedFrom['token'] = res.source!;
    }

    return {
        type: authType,
        providerName,
        username,
        email,
        password,
        token,
        headerName: 'Authorization',
        resolution: {
            resolvedFrom,
            tokenResolver: TokenResolverType.STATIC,
            isPlaceholder: false
        },
        resolverType: TokenResolverType.STATIC
    };
}
